### DISCUSSION TITLE
Is it possible to render the code in markdown instead of html ?

### DISCUSSION THREAD

**Discussion by emiliengarreau** [mkdocstrings/mkdocstrings#346] [Q&A] [answered]
I'd like to know if it is possible to render the generated documentation in markdown instead of html.

To give a brief overview of the context I'd like to generate the full documentation of a monorepo using the [spotify plugin](https://github.com/backstage/mkdocs-monorepo-plugin). A monorepo equal multiple library requirements, languages,... Hence, I need to be able to render my generated markdown before running the spotify plugin.

I guess using mkdocstring as a library could allow me to perform this rendering.
If you have other ideas I am all ears :).

And most of all thankssss for the great tool  :+1: :heart:

**Comment by pawamoy** [comment]
I'm not sure to follow: is there an issue between mkdocstrings and the monorepo plugin, making them incompatible?

**Reply by emiliengarreau** [reply]
Not at all. I wasn't clear :).

The issue is that the final docker which will build the final documentation  won't have access to any of the libraries requirements.

Example:

```
docs
libs
 ├─lib1
 │  ├─Dockerfile
 │  └─docs
 │     ├─template.md
 │     └─autogen_markdown.py
 └─lib2
    ├─Dockerfile
    └─docs
       ├─template.md
       └─autogen_markdown.py
mkdocs.yml
```

1. docker lib1 build its doc (render to markdown)
2. docker lib2 build its doc (render to markdown)
3. global docker without librariries dependencies create the full documentation website (render to html)

So my question is: Is there a way to create this autogen_markdown.py using Mkdocstring (a bit like [keras-doc](https://github.com/keras-team/keras-autodoc))?

Is it clearer ?

**Reply by pawamoy** [reply]
Yes it's more clear now, thanks :slightly_smiling_face: 
Unfortunately the answer is no, there is no way to do that. mkdocstrings is a plugin, and therefore is integrated into the MkDocs process of rendering markdown pages to a static HTML website. There's no intermediary markdown output, because autodoc instructions *are* markdown, and transformed into HTML while converting markdown to HTML.

What I fail to understand is: how is the global docker image supposed to render everything to HTML without libraries? You mean, without MkDocs plugins? But then when is the mkdocs-monorepo-plugin used, if not in the global docker image?

**Reply by emiliengarreau** [reply]
The idea was to used mkdocstring as a library inside the autogen_markdown.py.

> What I fail to understand is: how is the global docker image supposed to render everything to HTML without libraries? You mean, without MkDocs plugins? But then when is the mkdocs-monorepo-plugin used, if not in the global docker image?

I meant without the requirements from lib1 and lib2.
For example lib1 has numpy==2.N requirement and lib2 has numpy==1.1 requirement.

Example:

- In the beginning there was:

My awsome template.md

```
# Reference
::: lib1.my_class_A
```
 - Fill the template.md with autogen_markdown.py

```
# Reference

An awesome docstring explaining how works my_class_A
```

- Do that for all the libraries

- Since we have all the generated markdown, we don't need to have access to the libraries requirements anymore.

- Build all the monorepo documentation 

Is it clearer ?

**Reply by pawamoy** [reply]
Clearer again, thanks :smile: 
Maybe [mocking libraries](https://mkdocstrings.github.io/handlers/python/#mocking-libraries) could help?
Otherwise maybe you could hack the MkDocs process (patches? plugin?) to output the page contents after Markdown is converted to HTML to a temporary location. Indeed, HTML is Markdown (or rather, Markdown is a superset of HTML), so the monorepo plugin would have no trouble rendering it a second time (it should have no effect, though not sure of that depending on the loaded markdown extensions).

**Reply by emiliengarreau** [reply]
Oh yes you are right I could only convert the html files to .md and makes the monorepo find them easily.
I'll try some hacks :)

thanks :)

**Comment by emiliengarreau** [comment, marked as answer]
Thanks @pawamoy I did like you told me :).

I created a simple plugins which dumps html to markdown :).

```python
from mkdocs.plugins import BasePlugin
from pathlib import Path
from mkdocs.config.config_options import Type
from typing import Tuple


class DumpMarkdownPlugin(BasePlugin):
    config_scheme: Tuple[Tuple[str, Type]] = (("target_folder", Type(str, default="rendered_docs")),)

    def on_page_content(self, html: str, page, config, files):
        target_folder = Path(self.config["target_folder"])
        rendered_md = target_folder / page.file.src_path
        rendered_md.parent.mkdir(parents=True, exist_ok=True)
        with rendered_md.open("w") as f:
            f.write(html)
        return html

```

**Reply by pawamoy** [reply]
My sponsors might now want to look at https://mkdocstrings.github.io/griffe2md/ :slightly_smiling_face:
